File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def isValid(root, left, right):
            if not root:
                return True

            if root.val < right and root.val > left:
                lValid = isValid(root.left, left, root.val)
                rValid = isValid(root.right, root.val, right)
                return lValid and rValid

            return False

        return isValid(root, -float("inf"), float("inf"))

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        int[] count = new int[1];
        counter(root, count, root.val);
        return count[0];
    }

    public void counter(TreeNode root, int[] count, int largest) {
        if (root == null) {
            return;
        }

        if (root.val >= largest) {
            count[0] += 1;
        }

        largest = Math.max(root.val, largest);

        counter(root.left, count, largest);
        counter(root.right, count, largest);

    }
}
